src/storage/cache/rpc/mod.rs

```diff
@@ -0,0 +1,375 @@
+
+                    // File block request
+                    // Submit the handler to the worker pool
+                    // When the handler is done, send the response to the done channel
+                    // Response need to contain the response header and body
+                    let handler = FileBlockHandler::new(req_header, req_body, done_tx.clone());
+                    if let Ok(()) = self
+                        .worker_pool
+                        .try_submit_job(Box::new(handler))
+                        .map_err(|err| {
+                            debug!("Failed to submit job: {:?}", err);
+                        })
+                    {
+                        debug!("Submitted job to worker pool");
+                    }
+                } else {
+                    debug!(
+                        "FileBlockRpcServerHandler: Inner request type is not matched: {:?}",
+                        req_header.op
+                    );
+                }
+            }
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct TestFilePacket {
+        pub seq: u64,
+        pub op: u8,
+        pub status: PacketStatus,
+        pub timestamp: u64,
+        pub buffer: BytesMut,
+    }
+
+    impl TestFilePacket {
+        pub fn new(block_request: &FileBlockRequest) -> Self {
+            let block_request_buf = block_request.encode();
+            let mut buffer = BytesMut::new();
+            buffer.extend_from_slice(block_request_buf.as_slice());
+            Self {
+                // Will be auto set by client
+                seq: 0,
+                // Set operation
+                op: ReqType::FileBlockRequest.to_u8(),
+                status: PacketStatus::Pending,
+                timestamp: 0,
+                buffer,
+            }
+        }
+    }
+
+    impl Packet for TestFilePacket {
+        fn seq(&self) -> u64 {
+            self.seq
+        }
+
+        fn set_seq(&mut self, seq: u64) {
+            self.seq = seq;
+        }
+
+        fn op(&self) -> u8 {
+            self.op
+        }
+
+        fn set_op(&mut self, op: u8) {
+            self.op = op;
+        }
+
+        fn set_timestamp(&mut self, timestamp: u64) {
+            self.timestamp = timestamp;
+        }
+
+        fn get_timestamp(&self) -> u64 {
+            self.timestamp
+        }
+
+        fn set_req_data(&mut self, _data: &[u8]) -> Result<(), RpcError<String>> {
+            // Try to set the request data
+            Ok(())
+        }
+
+        fn get_req_data(&self) -> Result<Vec<u8>, RpcError<String>> {
+            // Try to get the request data
+            // check current buffer is req
+            let req = FileBlockRequest::decode(&self.buffer)?;
+
+            // Return a 4MB vec
+            Ok(req.encode())
+        }
+
+        fn set_resp_data(&mut self, data: &[u8]) -> Result<(), RpcError<String>> {
+            // Try to set the response data
+            self.buffer.clear();
```

Comment:
Why not decode directly and store the structured data, instead of [u8] ?

Reply:
We can do decode inside set_resp_data or set_result, passing in buffer here just avoids putting the decode logic into the rpc (decode requires type).